### PR TITLE
Https Transports

### PR DESCRIPTION
**Pull request by gbaz**
At this point Curl is working both ways, powershell appears working both ways, and wget has download but not upload. Etag support should be added back, but this does not seem to me a deal breaker.

### CODE REVIEW

**Comment by gbaz** [comment]
Note that the "uriToSecure" function in HttpUtils.hs converts all http:// to https:// urls, so as to upgrade users with old urls in their config.

I think that this is perhaps overeager and we might want to do something more clever here...

File: cabal-install/Distribution/Client/BuildReports/Upload.hs [hunk 1]
```diff
@@ -5,17 +5,17 @@ module Distribution.Client.BuildReports.Upload
     ( BuildLog
     , BuildReportId
     , uploadReports
-    , postBuildReport
-    , putBuildLog
     ) where
 
+{-
```
Comment (tibbe) [inline cabal-install/Distribution/Client/BuildReports/Upload.hs]: Why is this commented out? Work in progress?

**Comment by gbaz** [comment]
I commented out a most of the BuildReport uploading, actually. @dcoutts assured me that it currently was A) not turned on and B) not working, so there was no harm in rendering it busted for the time being. I'm fine with working further on it, but it is hard to feel a lot of motivation when you know that it wasn't working or usable to begin with, and when it does become usable, it will probably get rewritten anyway.

**Comment by dcoutts** [comment]
Yeah, that's ok, uploading build reports isn't working.

**Comment by 23Skidoo** [comment]
Sorry that I'm slow with code reviews - got sick, will hopefully recover in a few days.

**Comment by dcoutts** [comment]
Going through this now. I'm going to rebase and combine several of the patches so it's a bit clearer to see as a final set of changes, rather than a development trail. May also add some additional refactoring on top.

**Comment by dcoutts** [comment]
Ok, I'm planning to make one change on top (apart from a little refactoring).

The code assumes that we can always use an https address, even if the repo is specified as http. I don't think we can make that assumption in general, though it is true for the community hackage.h.o. So here's my plan, we keep track of whether the HttpTransport can do HTTPS (easy) and we keep track of whether the repository uses https or is one where we can/should transparently upgrade to https. The case in which we want to complain is then a little tricky, we need to know that we didn't deliberately configure a plain-http transport, and that the server can support both http and https. So the extra info we need is to keep track of whether the user specifically selected a transport or if it was picked automatically.

So I think I might build this bit on top of #2627 because that'll make it easier to keep extra info about particular repos.

**Comment by gbaz** [comment]
That additional change sounds like it makes sense to me.

**Comment by dcoutts** [comment]
Ok, I've got my changes ready as #2687

I've done a reasonable amount of testing of intermediate states while hacking on this. But I suggest we dogfood it a bit to check it's all working as expected. In particular needs checking on windows, and the support for getting etags and error messages needs checking on windows.

**Comment by 23Skidoo** [comment]
Closing in favour of #2687.

**Pull request closed without merging**